controllers/openshift_controller.go

```diff
@@ -1035,9 +993,75 @@ func (r *KataConfigOpenShiftReconciler) processKataConfigInstallRequest() (ctrl.
 		r.Log.Info("SCNodeRole is: " + machinePool)
 	}
 
-	wasMcJustCreated, err := r.createExtensionMc(machinePool)
-	if err != nil {
-		return ctrl.Result{Requeue: true}, nil
+	// Based on LayeredImageDeployment feature, we'll either create an ExtensionMachineConfig or
+	// just apply the image
+	wasMcJustCreated := false
+
+	// If featuregate LayeredImageDeployment is enabled then create the image based MC
+	// If it's disabled, create the extension based MC
+
+	if r.FeatureGatesStatus[featuregates.LayeredImageDeployment] {
+		r.Log.Info("PreviousMcName and PreviousMethod, ImageBasedDeployment=true",
+			"PreviousMcName", r.deploymentState.PreviousMcName, "PreviousMethod", r.deploymentState.PreviousMethod)
+		// Check if LayeredImageDeployment config is provided
+		// If not log an error and return
+		if !r.FeatureGates.IsFeatureConfigMapPresent(context.TODO(), featuregates.LayeredImageDeployment) {
+			r.Log.Info("LayeredImageDeployment feature is enabled but no configuration provided")
+			err := fmt.Errorf("LayeredImageDeployment feature is enabled but no configuration provided")
+			return ctrl.Result{Requeue: true}, err
+		}
+
+		// Check if the previous deployment method was different
+		// When we first time enter this condition the PreviousMethod will be empty
+		if r.deploymentState.PreviousMethod != "image" {
+			// Rollback the previous deployment
+			if r.deploymentState.PreviousMcName != "" {
+				// Delete the previous MachineConfig
+				err := r.deleteMc(r.deploymentState.PreviousMcName)
+				if err != nil {
+					r.Log.Info("Error deleting previous MachineConfig", "MachineConfig",
+						r.deploymentState.PreviousMcName, "err", err)
+					return ctrl.Result{Requeue: true}, err
+				}
+			}
+			wasMcJustCreated, err = r.createMc(machinePool)
+			if err != nil {
+				r.Log.Info("Error creating image MachineConfig for Image based deployment", "err", err)
+				return ctrl.Result{Requeue: true}, err
+			}
```

Comment:
Also, is it guaranteed that if we delete the previous MC and create the new one right away, the MCO will handle this correctly and sequence the operations properly, is the result well-defined?  If not or if we can't tell we'd better return and reschedule after the deletion and only create the new MC once we can verify that the operation triggered by the deletion of the previous one has finished.